src/python/pants/source/source_root_test.py

```diff
@@ -1,44 +1,157 @@
 # Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+import os
 from pathlib import PurePath
+from typing import Iterable, Optional, cast
 
 import pytest
 
-from pants.source.source_root import NoSourceRootError, SourceRootPatternMatcher
+from pants.engine.fs import Digest, PathGlobs, Snapshot
+from pants.source.source_root import (
+    OptionalSourceRoot,
+    SourceRootConfig,
+    SourceRootRequest,
+    get_source_root,
+)
+from pants.testutil.engine.util import MockGet, create_subsystem, run_rule
+
+
+def _find_root(
+    path: str,
+    patterns: Iterable[str],
+    marker_filenames: Optional[Iterable[str]] = None,
+    existing_marker_files: Optional[Iterable[str]] = None,
+) -> Optional[str]:
+    source_root_config = create_subsystem(
+        SourceRootConfig,
+        root_patterns=list(patterns),
+        marker_filenames=list(marker_filenames or []),
+    )
+    # This inner function is passed as the callable to the mock, to allow recursion in the rule.
+    def _mock_fs_check(pathglobs: PathGlobs) -> Snapshot:
+        for glob in pathglobs.globs:
+            if glob in (existing_marker_files or []):
+                d, f = os.path.split(pathglobs.globs[0])
+                return Snapshot(digest=Digest("111", 111), files=(f,), dirs=(d,))
+        return Snapshot(digest=Digest("000", 000), files=tuple(), dirs=tuple())
+
+    def _do_find_root(src_root_req: SourceRootRequest) -> OptionalSourceRoot:
+        return cast(
+            OptionalSourceRoot,
+            run_rule(
+                get_source_root,
+                rule_args=[src_root_req, source_root_config],
+                mock_gets=[
+                    MockGet(
+                        product_type=OptionalSourceRoot,
+                        subject_type=SourceRootRequest,
+                        mock=_do_find_root,
+                    ),
+                    MockGet(product_type=Snapshot, subject_type=PathGlobs, mock=_mock_fs_check),
+                ],
+            ),
+        )
+
+    source_root = _do_find_root(SourceRootRequest(PurePath(path))).source_root
+    return None if source_root is None else source_root.path
 
 
 def test_source_root_at_buildroot() -> None:
-    srpm = SourceRootPatternMatcher(("/",))
-    assert PurePath(".") == srpm.find_root("foo/bar.py")
-    assert PurePath(".") == srpm.find_root("foo/")
-    assert PurePath(".") == srpm.find_root("foo")
-    with pytest.raises(NoSourceRootError):
-        srpm.find_root("../foo/bar.py")
+    def find_root(path):
+        return _find_root(path, ("/",))
+
+    assert "." == find_root("foo/bar.py")
+    assert "." == find_root("foo/")
+    assert "." == find_root("foo")
+    with pytest.raises(ValueError):
```

Comment:
there should be a match param here to make sure we are getting the ValueError we are expecting and not a different one.